File: nn/js/matrix.js
// ----------------------------------------------------------------------------------//
// Matrix Library for use with Neural Networks
// - from Daniel Shiffman's youtube playlist (10)
// AI Samples (( v0.1 ))
// CRKS | September 1, 2018
// ----------------------------------------------------------------------------------//

// matrix definition
class Matrix {
  constructor(rows, cols) {
    this.rows = rows;
    this.cols = cols;
    this.data = [];

     for (let i = 0; i < this.rows; i++) {
       this.data[i] = [];
       for (let j = 0; j < this.cols; j++) {
         this.data[i][j] = 0;
       }
     }
  }

  static fromArray(array) {
    let matrix = new Matrix(array.length, 1);
    for (let i = 0; i < array.length; i++) {
      matrix.data[i][0] = array[i];
    }
    return matrix;
  }

  toArray() {
    let array = [];
    for (let i = 0; i < this.rows; i++) {
      for (let j = 0; j < this.cols; j++) {
        array.push(this.data[i][j]);
      }
    }
    return array;
  }

  static subtract(a, b) {
    let result = new Matrix(a.rows, a.cols);
    for (let i = 0; i < result.rows; i++) {
      for (let j = 0; j < result.cols; j++) {
        result.data[i][j] = a.data[i][j] - b.data[i][j];
      }
    }
    return result;
  }

  static transpose(a) {
    let result = new Matrix(a.cols, a.rows);
    for (let i = 0; i < a.rows; i++) {
      for (let j = 0; j < a.cols; j++) {
        result.data[j][i] = a.data[i][j];
      }
    }
    return result;
  }

  static multiply(a, b) {
    if (a.cols !== b.rows) {
      console.log('Columns of A must match rows of B!')
      return undefined;
    }
    let result = new Matrix(a.rows, b.cols);
    for (let i = 0; i < result.rows; i++) {
      for (let j = 0; j < result.cols; j++) {
        let sum = 0;
        for (let k = 0; k < a.cols; k++) {
          sum += a.data[i][k] * b.data[k][j];
        }
        result.data[i][j] = sum;
      }
    }
    return result;
  }

  randomize(n) {
    for (let i = 0; i < this.rows; i++) {
      for (let j = 0; j < this.cols; j++) {
        this.data[i][j] = Math.random() * 2 - 1;
      }
    }
  }

  map(func) {
    // apply a function to every value in the matrix
    for (let i = 0; i < this.rows; i++) {
      for (let j = 0; j < this.cols; j++) {
        let val = this.data[i][j];
        this.data[i][j] = func(val);
      }
    }
    return this;
  }

  static map(m, func) {
    // apply a function to every value in the matrix
    let result = new Matrix(m.rows, m.cols);
    for (let i = 0; i < m.rows; i++) {
      for (let j = 0; j < m.cols; j++) {
        let val = m.data[i][j];
        result.data[i][j] = func(val);
      }
    }
    return result;
  }

  add(n) {
    if (n instanceof Matrix) {
      for (let i = 0; i < this.rows; i++) {
        for (let j = 0; j < this.cols; j++) {
          this.data[i][j] += n.data[i][j];
        }
      }
    } else {
      for (let i = 0; i < this.rows; i++) {
        for (let j = 0; j < this.cols; j++) {
          this.data[i][j] += n;
        }
      }
    }
  }

  multiply(n) {
    if (n instanceof Matrix) {
      if (this.rows !== n.rows || this.cols !== n.cols) {
        console.log('Columns and Rows of A must match Columns and Rows of B.');
        return;
      }

      // hadamard product
      for (let i = 0; i < this.rows; i++) {
        for (let j = 0; j < this.cols; j++) {
          this.data[i][j] *= n.data[i][j];
        }
      }
    } else {
      // Scalar product
      return this.map(e => e * n);
    }
  }

  print() {
    for (let i = 0; i < this.rows; i++) {
      console.log(this.data[i]);
    }
  }
}

test = new Matrix(4,4);
test.randomize();
test.print();

File: nn/py/matrix.py
# ----------------------------------------------------------------------------------#
# Matrix Library for use with Neural Networks
# AI Samples (( v0.1 ))
# CRKS | September 4, 2018
# ----------------------------------------------------------------------------------#
import random

# matrix definition
class Matrix:
    # initialize the matrix
    def __init__(self, rows, cols):
        self.rows = rows
        self.cols = cols
        self.data = [[0 for i in range(cols)] for j in range(rows)]

    # print - prints square matrix to console
    def Print(self):
        for i in range(self.rows):
            print(self.data[i])

    # map - used to apply general functions to each matrix value
    def map(self,function):
        for i in range(self.rows):
            for j in range(self.cols):
                value = self.data[i][j]
                self.data[i][j] = function(value, i, j)

    # randomize - randomly sets all matrix values between -1 and 1
    def randomize(self):
        def rand(x, i, j):
            x = random.uniform(-1,1)
            return x
        self.map(rand)

    # add - if given a single digit, applies scalar addition to matrix
    #     - if given a matrix, applies element-wise addition
    def add(self,n):
        if isinstance(n, Matrix):
            if n.cols != self.cols or n.rows != self.rows:
                print('ERROR: Matrices must be of the same dimensions.')
            else:
                def add(x, i, j):
                    x = x + n.data[i][j]
                    return x
                self.map(add)
        else:
            def add(x, i, j):
                x = x + n
                return x
            self.map(add)

    # scale / multiply - takes the scalar product or hadmard depending on input
    def scale(self,n):
        if isinstance(n, Matrix):
            if n.cols != self.cols or n.rows != self.rows:
                print('ERROR: Matrices must be of the same dimensions.')
            else:
                def scale(x, i, j): # hadamard product
                    x = x * n.data[i][j]
                    return x
                self.map(scale)
        else:
            def scale(x, i, j): # scalar multiplication
                x = x * n
                return x
            self.map(scale)

    # subtract - STATIC METHOD
    @staticmethod
    def subtract(a,b):
        result = Matrix(a.rows, a.cols)
        if isinstance(b, Matrix):
            if a.cols != b.cols or a.rows != b.rows:
                print('ERROR: Matrices must be of the same dimensions.')
            else:
                def subtract(x, i, j):
                    x = x - b.data[i][j]
                    return x
                result.map(subtract)
        else:
            result = Matrix(a.rows, a.cols)
            def subtract(x, i, j):
                x = x - b
                return x
            result.map(subtract)
        return result

    # multiply - matrix dot? multiplication on inputs, outputs new matrix
    @staticmethod
    def multiply(a,b):
        if a.cols != b.rows:
            print('ERROR: Matrices must of the correct dimensions.')
        else:
            result = Matrix(a.rows,b.cols)
            def multiply(x,i,j):
                sum = 0
                for k in range(a.cols):
                    sum = sum + (a.data[i][k] * b.data[k][j])
                return sum
            result.map(multiply)
            return result

    # mapboth - static mapping method
    @staticmethod
    def mapboth(matrix,function):
        result = Matrix(matrix.rows, matrix.cols)
        def mapboth(x, i, j):
            x = function(matrix.data[i][j], i, j)
            return x
        result.map(mapboth)
        return result

    # transpose  matrix
    @staticmethod
    def transpose(matrix):
        result = Matrix(matrix.cols, matrix.rows)
        def transpose(x, i, j):
            x = matrix.data[j][i]
            return x
        result.map(transpose)
        return result

    @staticmethod
    def fromArray(array):
        result = Matrix(len(array), 1)
        for i in range(len(array)):
            result.data[i][0] = array[i]
        return result

    @staticmethod
    def toArray(matrix):
        result = []
        for i in range(matrix.rows):
            for j in range(matrix.cols):
                result.append(matrix.data[i][j])
        return result
